de.marw.cdt.cmake.core/src/main/java/de/marw/cdt/cmake/core/internal/BuildscriptGenerator.java

```diff
@@ -114,13 +119,150 @@ public void initialize(IProject project, IManagedBuildInfo info, IProgressMonito
     config = info.getDefaultConfiguration();
     builder = config.getEditableBuilder();
   }
+  
+  public static void createBuildFolderIfNeeded(final IProject project, CMakePreferences prefs, IConfiguration config) {
+    if(prefs == null) {
+      // fallback to the default values
+      prefs = new CMakePreferences();
+    }
+    
+    // resolve variables in the path
+    final IPath buildDestination;
+    try {
+      ICdtVariableManager mngr = CCorePlugin.getDefault().getCdtVariableManager();
+      String buildPathString = mngr.resolveValue(prefs.getBuildDirectory(), "", "",
+          ManagedBuildManager.getDescriptionForConfiguration(config));
+      buildDestination = new Path(buildPathString);
+    } catch (CdtVariableException e) {
+      createConfigMarker(project, IMarker.SEVERITY_ERROR, "Unable to resolve build path: ", e);
+      return;
+    }
+    
+    // clear existing markers
+    try {
+      project.deleteMarkers(MARKER_ID, true, IResource.DEPTH_ZERO);
+    } catch (CoreException e) {
+      // no marker needs to be deleted
+    }
+    
+    // abort if the destination path is not absolute
+    if (!buildDestination.isAbsolute()) {
+      return;
+    }
+    
+    final IWorkspace workspace = project.getWorkspace();
+    final IPathVariableManager pathMan = workspace.getPathVariableManager();
+    // use a pseudo unique hashcode for each project to avoid errors when creating
```

Comment:
Sounds hacky to me. Is it really needed to add a variable per project to the platform?
All those variables added here will show up in the UI, I guess.
Removing old build directories should be handled when the uses changes the dir and clicks apply, IMHO.

Reply:
I just read that the Path Variables are [bound to `IResource`s](https://help.eclipse.org/photon/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2FresAdv_variables.htm&cp=2_0_11_7), so I should be able to remove the hashcode by getting the `IPathVariableManager` from `project`. Only one variable should show up in its own project. But I suggest using a variable to remember the last build directory (a hacky alternative would be to save the path in the preferences before apply is clicked) and remove it if necessary.

Old build directories are only removed when the user clicks apply or in the special case that the new build directory does not exist before building (e.g. when someone modifies .cproject files).

```
final IWorkspace workspace = project.getWorkspace();
final IPathVariableManager pathMan = project.getPathVariableManager();
final String symLinkName = "CMAKE_BUILD_DIR";
```
This solution would be way cleaner.

Reply:
Why not use a Path Variable bound to the *workspace*? For the build folder, users can set it to e.g. '${WorkSpaceVarPointingToOurCommonOutputRoot}/${ProJName}/${ConfigName}'.
To re-create a linked output folder in case a user deleted it, we just need to store the information that the folder is a *linked* folder in the project preferences.